src/generic/trilinos_eigen_solver.h

```diff
@@ -538,10 +538,91 @@ namespace oomph
       // This is odd and not expected, examine carefully
       DoubleVector Y(x.distribution_pt());
       Linear_solver_pt->solve(AsigmaM_pt, X, Y);
-      // Need to synchronise
-      //#ifdef OOMPH_HAS_MPI
-      //   Problem_pt->synchronise_all_dofs();
-      //#endif
+
+      for (unsigned i = 0; i < n_row_local; i++)
+      {
+        y(0, i) = Y[i];
+      }
+
+      // Now loop over the others and resolve
+      for (unsigned v = 1; v < n_vec; ++v)
+      {
+        M_pt->multiply(x.doublevector(v), X);
+        Linear_solver_pt->resolve(X, Y);
+
+        for (unsigned i = 0; i < n_row_local; i++)
+        {
+          y(v, i) = Y[i];
+        }
+      }
+    }
+  };
+
+
+  //====================================================================
+  /// Class for the adjoing problem shift invert operation
+  //====================================================================
+  class AdjointProblemBasedShiftInvertOperator
+    : public DoubleMultiVectorOperator
+  {
+  private:
+    // Pointer to the problem
+    Problem* Problem_pt;
+
+    // Pointer to the linear solver used
+    LinearSolver* Linear_solver_pt;
+
+    // Storage for the shift
+    double Sigma;
+
+    // Storage for the matrices
+    CRDoubleMatrix *M_pt, *AsigmaM_pt;
+
+  public:
+    AdjointProblemBasedShiftInvertOperator(
+      Problem* const& problem_pt,
+      LinearSolver* const& linear_solver_pt,
+      const double& sigma = 0.0)
+      : Problem_pt(problem_pt), Linear_solver_pt(linear_solver_pt), Sigma(sigma)
+    {
+      // Before we get into the Arnoldi loop solve the shifted matrix problem
+      // Allocated Row compressed matrices for the mass matrix and shifted main
+      // matrix
+      M_pt = new CRDoubleMatrix(problem_pt->dof_distribution_pt());
+      AsigmaM_pt = new CRDoubleMatrix(problem_pt->dof_distribution_pt());
+
+      // Assemble the matrices
+      problem_pt->get_eigenproblem_matrices(*M_pt, *AsigmaM_pt, Sigma);
+
+      // Get the transpose of the mass and jacobian
+      // NB Only difference to ProblemBasedShiftInvertOperator
+      M_pt->get_matrix_transpose(M_pt);
+      AsigmaM_pt->get_matrix_transpose(AsigmaM_pt);
+
+      // Do not report the time taken
+      Linear_solver_pt->disable_doc_time();
+    }
+
+
+    // Now specify how to apply the operator
+    void apply(const DoubleMultiVector& x, DoubleMultiVector& y) const
+    {
+      const unsigned n_vec = x.nvector();
+      const unsigned n_row_local = x.nrow_local();
+      if (n_vec > 1)
+      {
+        Linear_solver_pt->enable_resolve();
+      }
+      // Solve the first vector
+
+      DoubleVector X(x.distribution_pt());
+      // Premultiply by mass matrix
+      M_pt->multiply(x.doublevector(0), X);
+      // For some reason I need to create a new vector and copy here
+      // This is odd and not expected, examine carefully
```

Comment:
Indeed... ("examine carefully" -- what exactly is the problem? One for the oopmh-lunch?)

Reply:
Ah. This is an artefact from copying from the non-adjoint version. I'm not sure whether to keep this comment or refer to the earlier one, or indeed figure out what is the problem and fix both!

Reply:
We discussed this in oomph-lunch. After examining carefully, the fix is to replace the comment with `Create output vector`.

Reply:
Done